### PR TITLE
p2p: Headers-sync followups

### PR DESCRIPTION
**Pull request by sdaftuar**
Remove BCLog::HEADERSSYNC and move all headerssync logging to BCLog::NET.

Bypass headers anti-DoS checks for NoBan peers

Also fix a typo that was introduced in PR25717.

### CODE REVIEW

**Comment by sdaftuar** [comment]
@ajtowns @instagibbs These should address the remaining review items that came up in #25717.

File: test/functional/p2p_headers_sync_with_minchainwork.py [hunk 1]
```diff
@@ -40,17 +40,32 @@ def setup_network(self):
...
 
     def test_chains_sync_when_long_enough(self):
         self.log.info("Generate blocks on the node with no required chainwork, and verify nodes 1 and 2 have no new headers in their headers tree")
         with self.nodes[1].assert_debug_log(expected_msgs=["[net] Ignoring low-work chain (height=14)"]), self.nodes[2].assert_debug_log(expected_msgs=["[net] Ignoring low-work chain (height=14)"]):
             self.generate(self.nodes[0], NODE1_BLOCKS_REQUIRED-1, sync_fun=self.no_op)
 
-        for node in self.nodes[1:]:
+        # Node3 should always allow headers due to noban permissions
+        self.log.info("Check that node3 will sync headers (due to noban permissions)")
+        def check_node3_chaintips(num_tips, tip_hash, height):
```
Comment (fanquake) [inline test/functional/p2p_headers_sync_with_minchainwork.py:57]:
```bash
/tmp/cirrus-ci-build/test/functional/p2p_headers_sync_with_minchainwork.py:57: unused variable 'num_tips' (100% confidence)
test/functional/p2p_headers_sync_with_minchainwork.py:57:9: E306 expected 1 blank line before a nested definition, found 0
```

**Comment by jonatack** [comment]
<details><summary>Some initial WIP follow-up suggestions on first look-over of #25717.</summary><p>

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 3b97c9252e..96bace65d6 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -632,15 +632,15 @@ private:
      *              acceptance by the caller).
      */
     bool IsContinuationOfLowWorkHeadersSync(Peer& peer, CNode& pfrom,
-            std::vector<CBlockHeader>& headers)
+                                            std::vector<CBlockHeader>& headers)
         EXCLUSIVE_LOCKS_REQUIRED(peer.m_headers_sync_mutex, !m_headers_presync_mutex);
     /** Check work on a headers chain to be processed, and if insufficient,
      * initiate our anti-DoS headers sync mechanism.
      *
-     * @param[in]   peer                The peer whose headers we're processing.
-     * @param[in]   pfrom               CNode of the peer
-     * @param[in]   chain_start_header  Where these headers connect in our index.
-     * @param[in,out]   headers             The headers to be processed.
+     * @param[in]     peer                The peer whose headers we're processing.
+     * @param[in]     pfrom               CNode of the peer
+     * @param[in]     chain_start_header  Where these headers connect in our index.
+     * @param[in,out] headers             The headers to be processed.
      *
      * @return      True if chain was low work and a headers sync was
      *              initiated (and headers will be empty after calling); false
@@ -1467,6 +1467,9 @@ void PeerManagerImpl::ReattemptInitialBroadcast(CScheduler& scheduler)
 
 void PeerManagerImpl::FinalizeNode(const CNode& node)
 {
+    AssertLockNotHeld(m_headers_presync_mutex);
     NodeId nodeid = node.GetId();
     int misbehavior{0};
     {
@@ -2447,6 +2450,9 @@ bool PeerManagerImpl::CheckHeadersAreContinuous(const std::vector<CBlockHeader>&
 
 bool PeerManagerImpl::IsContinuationOfLowWorkHeadersSync(Peer& peer, CNode& pfrom, std::vector<CBlockHeader>& headers)
 {
+    AssertLockNotHeld(m_headers_presync_mutex);
     if (peer.m_headers_sync) {
         auto result = peer.m_headers_sync->ProcessNextHeaders(headers, headers.size() == MAX_HEADERS_RESULTS);
         if (result.request_more) {
@@ -2460,13 +2466,12 @@ bool PeerManagerImpl::IsContinuationOfLowWorkHeadersSync(Peer& peer, CNode& pfro
                 // it may be possible to bypass this via compactblock
                 // processing, so check the result before logging just to be
                 // safe.
-                bool sent_getheaders = MaybeSendGetHeaders(pfrom, locator, peer);
-                if (sent_getheaders) {
+                if (MaybeSendGetHeaders(pfrom, locator, peer)) {
                     LogPrint(BCLog::NET, "more getheaders (from %s) to peer=%d\n",
-                            locator.vHave.front().ToString(), pfrom.GetId());
+                             locator.vHave.front().ToString(), pfrom.GetId());
                 } else {
                     LogPrint(BCLog::NET, "error sending next getheaders (from %s) to continue sync with peer=%d\n",
-                            locator.vHave.front().ToString(), pfrom.GetId());
+                             locator.vHave.front().ToString(), pfrom.GetId());
                 }
             }
         }
@@ -2733,6 +2738,9 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
                                             std::vector<CBlockHeader>&& headers,
                                             bool via_compact_block)
 {
+    AssertLockNotHeld(m_headers_presync_mutex);
     size_t nCount = headers.size();
 
     if (nCount == 0) {
@@ -3131,6 +3139,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                                      const std::chrono::microseconds time_received,
                                      const std::atomic<bool>& interruptMsgProc)
 {
+    AssertLockNotHeld(m_headers_presync_mutex);
     LogPrint(BCLog::NET, "received: %s (%u bytes) peer=%d\n", SanitizeString(msg_type), vRecv.size(), pfrom.GetId());
```
</p></details>

Feel free to ignore. I also have a branch ready at https://github.com/jonatack/bitcoin/tree/net-add-missing-thread-safety-runtime-assertions for all the net and net_processing code and mentioned above only the runtime assertions directly related to changes in #25717.

File: test/functional/p2p_headers_sync_with_minchainwork.py (hunk 1 above)
Comment (sdaftuar) [reply test/functional/p2p_headers_sync_with_minchainwork.py:58]: Thanks, fixed.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2820,6 +2820,13 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
         }
     }
 
+    // If our peer has NetPermissionFlags::NoBan privileges, then bypass our
+    // anti-DoS logic (this saves bandwidth when we connect to a trusted peer
+    // on startup).
+    if (pfrom.HasPermission(NetPermissionFlags::NoBan)) {
+        already_validated_work = true;
```
Comment (naumenkogs) [inline src/net_processing.cpp:2827]: nit: after this change`trustworthy_work` or something could be more accurate, although I'm not sure it actually benefits the readers.

**Comment by Sjors** [comment]
utACK 94af3e43e20aa00b18e7a3f6d0f5fe3ad9494d97

I did not test if `NoBan` peers indeed sync in a single pass now.

**Comment by ajtowns** [comment]
> The `download` permission might be more suitable for this

For what it's worth, I interpret `download` as being more in the other direction -- "this peer can download from me, even if I'm in IBD" rather than "I feel safe downloading from this peer", so it seemed less suitable to me. But :shrug:

File: src/logging.h [hunk 3]
```diff
@@ -65,7 +65,6 @@ namespace BCLog {
 #endif
         UTIL        = (1 << 25),
         BLOCKSTORE  = (1 << 26),
-        HEADERSSYNC = (1 << 27),
```
Comment (ajtowns) [inline src/logging.h:68]: You could make this a scripted diff:

```
-BEGIN VERIFY SCRIPT-
sed -i '/headerssync/Id' src/logging.h src/logging.cpp
sed -i 's/BCLog::HEADERSSYNC/BCLog::NET/g' $(git grep -l BCLog::HEADERSSYNC)
-END VERIFY SCRIPT-
```

**Comment by Sjors** [comment]
We should probably document what we mean by these permissions :-)

**Comment by ajtowns** [comment]
> We should probably document what we mean by these permissions :-)

They're documented in `bitcoin-cli help getpeerinfo`:

```
    "permissions" : [                     (json array) Any special permissions that have been granted to this peer
      "str",                              (string) bloomfilter (allow requesting BIP37 filtered blocks and transactions),
                                          noban (do not ban for misbehavior; implies download),
                                          forcerelay (relay transactions that are already in the mempool; implies relay),
                                          relay (relay even in -blocksonly mode, and unlimited transaction announcements),
                                          mempool (allow requesting BIP35 mempool contents),
                                          download (allow getheaders during IBD, no disconnect after maxuploadtarget limit),
                                          addr (responses to GETADDR avoid hitting the cache and contain random records with the most up-to-date info).
                                          
      ...
    ],
```

and in `bitcoind -help` under `-whitebind`.

**Comment by Sjors** [comment]
Ok, in that case I tend to agree `noban` is better. I guess `download` should have been called `upload` :-)

**Comment by ajtowns** [comment]
ACK 94af3e43e20aa00b18e7a3f6d0f5fe3ad9494d97

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25960/commits/94af3e43e20aa00b18e7a3f6d0f5fe3ad9494d97

**Comment by naumenkogs** [comment]
ACK 94af3e43e20aa00b18e7a3f6d0f5fe3ad9494d97

**Pull request merged**